### PR TITLE
Fix parallel vae

### PR DESCRIPTION
**Pull request by gty111**
related to #271

### CODE REVIEW

**Comment by gty111** [comment]
In this PR, also fix #282

File: xfuser/model_executor/models/transformers/transformer_flux.py [hunk 1]
```diff
@@ -109,21 +109,20 @@ def forward(
...
-            txt_ids = txt_ids[0]
-        if img_ids.ndim == 3:
-            logger.warning(
-                "Passing `img_ids` 3d torch.Tensor is deprecated."
-                "Please remove the batch dimension and pass it as a 2d torch Tensor"
-            )
-            img_ids = img_ids[0]
-
-        ids = torch.cat((txt_ids, img_ids), dim=0)
+        # if txt_ids.ndim == 3:
```
Comment (Eigensystem) [inline xfuser/model_executor/models/transformers/transformer_flux.py]: why delete the check

File: xfuser/model_executor/pipelines/pipeline_flux.py [hunk 2]
```diff
@@ -309,19 +310,30 @@ def __call__(
...
+            latents = (
+                latents / self.vae.config.scaling_factor
+            ) + self.vae.config.shift_factor
+            
+            image = self.vae.decode(latents, return_dict=False)[0]
+            return image
+        
+        if not output_type == "latent":
+            if get_runtime_state().runtime_config.use_parallel_vae:
+                latents = get_world_group().broadcast_latent(latents,get_runtime_state().runtime_config.dtype)
```
Comment (Eigensystem) [inline xfuser/model_executor/pipelines/pipeline_flux.py]: It's not very rigorous. If we enable dp/cfg，we can not broadcast like this.

File: xfuser/model_executor/pipelines/pipeline_pixart_sigma.py [hunk 3]
```diff
@@ -323,11 +324,23 @@ def __call__(
...
+        
+        def vae_decode(latents):
+            image = self.vae.decode(
+                latents / self.vae.config.scaling_factor, return_dict=False
+            )[0]
+            return image
+        
+        if not output_type == "latent":
+            if get_runtime_state().runtime_config.use_parallel_vae:
+                latents = get_world_group().broadcast_latent(latents,get_runtime_state().runtime_config.dtype)
```
Comment (Eigensystem) [inline xfuser/model_executor/pipelines/pipeline_pixart_sigma.py]: as above

File: xfuser/model_executor/pipelines/pipeline_stable_diffusion_3.py [hunk 4]
```diff
@@ -379,16 +380,26 @@ def __call__(
...
+        def vae_decode(latents):
+            latents = (
+                    latents / self.vae.config.scaling_factor
+                ) + self.vae.config.shift_factor
+            image = self.vae.decode(latents, return_dict=False)[0]
+            return image
+        
+        if not output_type == "latent":
+            if get_runtime_state().runtime_config.use_parallel_vae:
+                latents = get_world_group().broadcast_latent(latents,get_runtime_state().runtime_config.dtype)
```
Comment (Eigensystem) [inline xfuser/model_executor/pipelines/pipeline_stable_diffusion_3.py]: as above

**Review by Eigensystem — CHANGES_REQUESTED** [review]

File: xfuser/model_executor/models/transformers/transformer_flux.py (hunk 1 above)
Comment (gty111) [reply xfuser/model_executor/models/transformers/transformer_flux.py]: In diffuser v0.30.3, it supports passing three dimensions tensor. So we do not need to lower the dimension of the tensor.
https://github.com/huggingface/diffusers/blob/v0.30.3/src/diffusers/models/transformers/transformer_flux.py#L380

File: xfuser/model_executor/pipelines/pipeline_flux.py (hunk 2 above)
Comment (gty111) [reply xfuser/model_executor/pipelines/pipeline_flux.py]: In the following commit, fix this issue.

**Comment by gty111** [comment]
To consider DP, we add `gather_broadcast_latents` and `is_dp_last_group` in `xFuserPipelineBaseWrapper`. 
`gather_broadcast_latents`: 
1. identify dp last group 
2. gather latents from dp last group and concatenate the latents in the batch dim
3. broadcast latents 

Then use default group to complete parallel vae.

When using parallel vae, only one process need to process img after vae. So `is_dp_last_group` is to correctly identify the process which needs to do sth after vae.

@Eigensystem

File: xfuser/model_executor/pipelines/pipeline_hunyuandit.py [hunk 5]
```diff
@@ -454,12 +455,22 @@ def __call__(
...
+            )[0]
+            return image
+        
+        if not output_type == "latent":
+            if get_runtime_state().runtime_config.use_parallel_vae:
+                latents = self.gather_broadcast_latents(latents)
+                vae_decode(latents)
+            else:
+                if is_dp_last_group():
+                    vae_decode(latents)
```
Comment (Eigensystem) [inline xfuser/model_executor/pipelines/pipeline_hunyuandit.py]: miss return value of func `vae_decode`

**Review by Eigensystem — CHANGES_REQUESTED** [review]

**Comment by Eigensystem** [comment]
I think it is not a elegant implementation because you combined all the dp latents into one tensor instead of using different dp groups to process different latent. It will increase the communication load.

**Comment by gty111** [comment]
> I think it is not a elegant implementation because you combined all the dp latents into one tensor instead of using different dp groups to process different latent. It will increase the communication load.

It is indeed not the best implementation. Using different dp groups need to refactor DistVAE. Since parallel vae is an option, maybe we can first fix parallel vae in this PR, and further optimize it in the future.

**Review by Eigensystem — APPROVED** [review]
LGTM.
